src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+# Tree
+## Compaction
+
+Tree compaction runs to the sound of music!
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
```

Comment:
It's not obvious to me how this is incremental. Presumably we do some bounded amount of work per committed op, but from reading the description it's not ruled out that we might get a single op and then have to compact every level of the tree. Is there any bound on how many levels get compacted per op? Or are the compactions partial, and if so how do we guarantee that they always finish in time?

Reply:
We might compact every level on a single op, yes. The incremental nature is that each compaction (from level A to level A+1) is split up to take place across `lsm_batch_multiple` ops.

> Or are the compactions partial, and if so how do we guarantee that they always finish in time?

Yep. Right now we guarantee this at the last beat of each compaction, which will not return control until it finishes. I'll leave this documentation as-is for now — I'm currently working on fixing compaction scheduling to spread work more evenly between beats.

Reply:
So the maximum compaction stall is when the largest-but-one level compacts into the largest level, divided by `lsm_batch_multiple` ops? That would be ~1/2 the database in reads+writes over 4 ops, if I'm understanding this correctly.

Reply:
- Each table has the same maximum size, regardless of whether it is a level-0 table or a level-N table.
- The number of tables increases exponentially at each successive level.
- So the key range spanned by a table in level `X+1` is typically smaller than the key range spanned by a table in level `X`.

So what that is the worst case stall, but it would require a weird write workload to trigger — on average 1/8 of the level-B tables will overlap with the level-A table. With how our actual trees are indexed (using a timestamp or a UUID) it is probably impossible, but I haven't worked through it entirely. (Unless you chose UUIDs antagonistically :stuck_out_tongue:).

(Also, although we currently set `lsm_batch_multiple` to 4, it will be set higher in production.